File: apps/mobile/lib/models/catalog.dart
// Mirrors apps/frontend/src/lib/{catalogListItem,types}.ts — same shapes, since both clients hit
// the same apps/backend REST API.

sealed class CatalogListItem {
  final String id;
  final String name;

  const CatalogListItem({required this.id, required this.name});

  factory CatalogListItem.fromJson(Map<String, dynamic> json) {
    return json['kind'] == 'folder'
        ? CatalogFolderItem(
            id: json['id'] as String,
            name: json['name'] as String,
            thumbnailFileId: json['thumbnailFileId'] as String?,
          )
        : CatalogVideoItem(
            id: json['id'] as String,
            name: json['name'] as String,
            mimeType: json['mimeType'] as String,
          );
  }
}

class CatalogFolderItem extends CatalogListItem {
  final String? thumbnailFileId;
  const CatalogFolderItem({required super.id, required super.name, this.thumbnailFileId});
}

class CatalogVideoItem extends CatalogListItem {
  final String mimeType;
  const CatalogVideoItem({required super.id, required super.name, required this.mimeType});
}

class WatchProgress {
  final String fileId;
  final String parentFolderId;
  final double positionSeconds;
  final double durationSeconds;
  final bool completed;

  const WatchProgress({
    required this.fileId,
    required this.parentFolderId,
    required this.positionSeconds,
    required this.durationSeconds,
    required this.completed,
  });

  factory WatchProgress.fromJson(Map<String, dynamic> json) => WatchProgress(
    fileId: json['fileId'] as String,
    parentFolderId: json['parentFolderId'] as String,
    positionSeconds: (json['positionSeconds'] as num).toDouble(),
    durationSeconds: (json['durationSeconds'] as num).toDouble(),
    completed: json['completed'] as bool,
  );
}

class StreamTrack {
  final int index;
  final String? language;
  final String? title;
  final String codecName;

  const StreamTrack({required this.index, this.language, this.title, required this.codecName});

  factory StreamTrack.fromJson(Map<String, dynamic> json) => StreamTrack(
    index: json['index'] as int,
    language: json['language'] as String?,
    title: json['title'] as String?,
    codecName: json['codecName'] as String,
  );
}

/// One rung of the adjustable-quality ladder — mirrors apps/backend/src/lib/qualityLadder.ts.
/// Clients never hardcode the ladder; they just render whatever the backend's probe response says
/// is available for this particular video (filtered to below its own source resolution there).
class QualityOption {
  final String label;
  final int height;

  const QualityOption({required this.label, required this.height});

  factory QualityOption.fromJson(Map<String, dynamic> json) =>
      QualityOption(label: json['label'] as String, height: json['height'] as int);
}

class ProbeResult {
  final List<StreamTrack> audioTracks;
  final List<StreamTrack> subtitleTracks;
  final double? durationSeconds;
  final int? sourceHeight;
  final int? sourceWidth;
  final List<QualityOption> availableQualities;

  const ProbeResult({
    required this.audioTracks,
    required this.subtitleTracks,
    this.durationSeconds,
    this.sourceHeight,
    this.sourceWidth,
    this.availableQualities = const [],
  });

  factory ProbeResult.fromJson(Map<String, dynamic> json) => ProbeResult(
    audioTracks: (json['audioTracks'] as List).map((t) => StreamTrack.fromJson(t)).toList(),
    subtitleTracks: (json['subtitleTracks'] as List).map((t) => StreamTrack.fromJson(t)).toList(),
    durationSeconds: (json['durationSeconds'] as num?)?.toDouble(),
    sourceHeight: json['sourceHeight'] as int?,
    sourceWidth: json['sourceWidth'] as int?,
    availableQualities: json['availableQualities'] == null
        ? const []
        : (json['availableQualities'] as List).map((q) => QualityOption.fromJson(q)).toList(),
  );
}

class ConvertedSubtitle {
  final int index;
  final String? language;
  final String? title;
  final String vtt;

  const ConvertedSubtitle({required this.index, this.language, this.title, required this.vtt});

  factory ConvertedSubtitle.fromJson(Map<String, dynamic> json) => ConvertedSubtitle(
    index: json['index'] as int,
    language: json['language'] as String?,
    title: json['title'] as String?,
    vtt: json['vtt'] as String,
  );
}

class HeroData {
  final String title;
  final String badge;
  final String playHref;
  final String? infoHref;

  const HeroData({required this.title, required this.badge, required this.playHref, this.infoHref});

  factory HeroData.fromJson(Map<String, dynamic> json) => HeroData(
    title: json['title'] as String,
    badge: json['badge'] as String,
    playHref: json['playHref'] as String,
    infoHref: json['infoHref'] as String?,
  );
}

class ContinueWatchingEntry {
  final CatalogListItem item;
  final WatchProgress progress;
  const ContinueWatchingEntry({required this.item, required this.progress});

  factory ContinueWatchingEntry.fromJson(Map<String, dynamic> json) => ContinueWatchingEntry(
    item: CatalogListItem.fromJson(json['item'] as Map<String, dynamic>),
    progress: WatchProgress.fromJson(json['progress'] as Map<String, dynamic>),
  );
}

class LibraryResponse {
  final bool empty;
  final HeroData? hero;
  final List<ContinueWatchingEntry> continueWatching;
  final List<CatalogListItem> folders;
  final List<CatalogListItem> videos;

  const LibraryResponse({
    required this.empty,
    this.hero,
    required this.continueWatching,
    required this.folders,
    required this.videos,
  });

  factory LibraryResponse.fromJson(Map<String, dynamic> json) => LibraryResponse(
    empty: json['empty'] as bool,
    hero: json['hero'] == null ? null : HeroData.fromJson(json['hero'] as Map<String, dynamic>),
    continueWatching: (json['continueWatching'] as List)
        .map((e) => ContinueWatchingEntry.fromJson(e))
        .toList(),
    folders: (json['folders'] as List).map((e) => CatalogListItem.fromJson(e)).toList(),
    videos: (json['videos'] as List).map((e) => CatalogListItem.fromJson(e)).toList(),
  );
}

class Season {
  final String id;
  final String name;
  const Season({required this.id, required this.name});
  factory Season.fromJson(Map<String, dynamic> json) =>
      Season(id: json['id'] as String, name: json['name'] as String);
}

/// Covers both of the backend's folder response shapes (`mode: "flat" | "series"`) — Dart lacks
/// TypeScript's discriminated-union ergonomics without codegen, so this is one class with the
/// season-specific fields left null in flat mode, same as optional fields on a plain JS object.
class FolderResponse {
  final String mode;
  final String folderId;
  final String title;
  final String backHref;
  final String? thumbnailFileId;
  final List<CatalogListItem> items; // flat mode only
  final List<Season> seasons; // series mode only
  final String? selectedSeasonId;
  final List<CatalogListItem> seasonItems;
  final List<CatalogListItem> specialItems;
  final List<WatchProgress> progress;

  const FolderResponse({
    required this.mode,
    required this.folderId,
    required this.title,
    required this.backHref,
    this.thumbnailFileId,
    this.items = const [],
    this.seasons = const [],
    this.selectedSeasonId,
    this.seasonItems = const [],
    this.specialItems = const [],
    this.progress = const [],
  });

  bool get isSeries => mode == 'series';

  factory FolderResponse.fromJson(Map<String, dynamic> json) => FolderResponse(
    mode: json['mode'] as String,
    folderId: json['folderId'] as String,
    title: (json['title'] as String?) ?? '',
    backHref: json['backHref'] as String,
    thumbnailFileId: json['thumbnailFileId'] as String?,
    items: json['items'] == null
        ? const []
        : (json['items'] as List).map((e) => CatalogListItem.fromJson(e)).toList(),
    seasons: json['seasons'] == null
        ? const []
        : (json['seasons'] as List).map((e) => Season.fromJson(e)).toList(),
    selectedSeasonId: json['selectedSeasonId'] as String?,
    seasonItems: json['seasonItems'] == null
        ? const []
        : (json['seasonItems'] as List).map((e) => CatalogListItem.fromJson(e)).toList(),
    specialItems: json['specialItems'] == null
        ? const []
        : (json['specialItems'] as List).map((e) => CatalogListItem.fromJson(e)).toList(),
    progress: json['progress'] == null
        ? const []
        : (json['progress'] as List).map((e) => WatchProgress.fromJson(e)).toList(),
  );
}

class EpisodeRef {
  final String id;
  final String name;
  const EpisodeRef({required this.id, required this.name});
  factory EpisodeRef.fromJson(Map<String, dynamic> json) =>
      EpisodeRef(id: json['id'] as String, name: json['name'] as String);
}

class EpisodeProgress {
  final double positionSeconds;
  final bool completed;
  const EpisodeProgress({required this.positionSeconds, required this.completed});
  factory EpisodeProgress.fromJson(Map<String, dynamic> json) => EpisodeProgress(
    positionSeconds: (json['positionSeconds'] as num).toDouble(),
    completed: json['completed'] as bool,
  );
}

class VideoResponse {
  final String fileId;
  final String title;
  final String backHref;
  final String parentFolderId;
  final String? nextFileId;
  final String? previousFileId;
  final List<EpisodeRef> episodes;
  final Map<String, EpisodeProgress> progressByFileId;
  final double initialPositionSeconds;
  final bool initialCompleted;
  final String seekMode; // "native" | "restart" | "raw" (MKV — see MkvVideoPlayer)
  final double? durationSeconds;
  final List<ConvertedSubtitle> subtitles;
  final double? introStart;
  final double? introEnd;
  final double? outroStart;

  const VideoResponse({
    required this.fileId,
    required this.title,
    required this.backHref,
    required this.parentFolderId,
    this.nextFileId,
    this.previousFileId,
    required this.episodes,
    required this.progressByFileId,
    required this.initialPositionSeconds,
    required this.initialCompleted,
    required this.seekMode,
    this.durationSeconds,
    required this.subtitles,
    this.introStart,
    this.introEnd,
    this.outroStart,
  });

  bool get isNative => seekMode == 'native';
  bool get isRaw => seekMode == 'raw';

  factory VideoResponse.fromJson(Map<String, dynamic> json) => VideoResponse(
    fileId: json['fileId'] as String,
    title: (json['title'] as String?) ?? '',
    backHref: json['backHref'] as String,
    parentFolderId: json['parentFolderId'] as String,
    nextFileId: json['nextFileId'] as String?,
    previousFileId: json['previousFileId'] as String?,
    episodes: (json['episodes'] as List).map((e) => EpisodeRef.fromJson(e)).toList(),
    progressByFileId: (json['progressByFileId'] as Map<String, dynamic>).map(
      (k, v) => MapEntry(k, EpisodeProgress.fromJson(v as Map<String, dynamic>)),
    ),
    initialPositionSeconds: (json['initialPositionSeconds'] as num).toDouble(),
    initialCompleted: json['initialCompleted'] as bool,
    seekMode: json['seekMode'] as String,
    durationSeconds: (json['durationSeconds'] as num?)?.toDouble(),
    subtitles: (json['subtitles'] as List).map((e) => ConvertedSubtitle.fromJson(e)).toList(),
    introStart: (json['introStart'] as num?)?.toDouble(),
    introEnd: (json['introEnd'] as num?)?.toDouble(),
    outroStart: (json['outroStart'] as num?)?.toDouble(),
  );
}

File: apps/backend/src/routes/catalog.ts
import { Router } from "express";
import { env } from "../config/env";
import { requireAuth } from "../middleware/auth";
import { isAdminEmail } from "../lib/admin";
import { isNativelyPlayable, isMkv, type ConvertedSubtitle } from "../lib/drive";
import {
  getCatalogFolder,
  listPublishedFoldersByParent,
} from "../lib/catalogFolders";
import {
  getCatalogVideo,
  listPublishedVideosByParent,
} from "../lib/catalogVideos";
import { getProgress, getProgressForFolder, getRecentInProgress } from "../lib/progress";
import { listSubtitleSetsByIds } from "../lib/subtitleSets";
import { naturalSort, orderEpisodes, type CatalogListItem } from "../lib/catalogListItem";

const router = Router();

router.get("/library", requireAuth, async (req, res) => {
  const [folderDocs, videoDocs] = await Promise.all([
    listPublishedFoldersByParent(env.driveRootFolderId),
    listPublishedVideosByParent(env.driveRootFolderId),
  ]);

  const folders = naturalSort(
    folderDocs.map((f) => ({
      kind: "folder" as const,
      id: f._id,
      name: f.title!,
      thumbnailFileId: f.thumbnailFileId,
    })),
  );
  const videos = naturalSort(
    videoDocs.map((v) => ({ kind: "video" as const, id: v._id, name: v.title!, mimeType: v.mimeType })),
  );
  const items: (CatalogListItem & { curatedAt: Date | null })[] = [
    ...folderDocs.map((f) => ({
      kind: "folder" as const,
      id: f._id,
      name: f.title!,
      thumbnailFileId: f.thumbnailFileId,
      curatedAt: f.curatedAt,
    })),
    ...videoDocs.map((v) => ({
      kind: "video" as const,
      id: v._id,
      name: v.title!,
      mimeType: v.mimeType,
      curatedAt: v.curatedAt,
    })),
  ];

  if (items.length === 0) {
    res.json({ empty: true, folders: [], videos: [], hero: null, continueWatching: [] });
    return;
  }

  // Real "most recently curated" pick, not a fabricated hero slot.
  const hero = [...items].sort((a, b) => {
    const at = a.curatedAt ? new Date(a.curatedAt).getTime() : 0;
    const bt = b.curatedAt ? new Date(b.curatedAt).getTime() : 0;
    return bt - at;
  })[0];

  // A ready-to-use path (or null, when a folder simply has no thumbnail yet) rather than a raw id
  // + kind — keeps clients from needing to know the folder-vs-video branching this backend already
  // knows (matches heroPlayHref/heroInfoHref's own "hand back a usable href" convention).
  const heroThumbnailPath =
    hero.kind === "video" ? `/api/thumbnail/${hero.id}` : hero.thumbnailFileId ? `/api/thumbnail-folder/${hero.id}` : null;
  let heroPlayHref = `/folder/${hero.id}`;
  const heroInfoHref = hero.kind === "folder" ? `/folder/${hero.id}` : undefined;
  if (hero.kind === "video") {
    heroPlayHref = `/watch/${hero.id}`;
  } else {
    const episodesInFolder = orderEpisodes(
      (await listPublishedVideosByParent(hero.id)).map((v) => ({
        id: v._id,
        name: v.title!,
        episodeOrder: v.episodeOrder,
      })),
    );
    heroPlayHref = episodesInFolder[0] ? `/watch/${episodesInFolder[0].id}` : `/folder/${hero.id}`;
  }

  const continueWatching = await getRecentInProgress(req.authUser!.userId, 12);
  const continueWatchingItems = (
    await Promise.all(
      continueWatching.map(async (progress) => {
        const video = await getCatalogVideo(progress.fileId);
        if (!video || video.status !== "published") return null; // unpublished or deleted — skip rather than break the row
        return {
          item: { kind: "video" as const, id: video._id, name: video.title!, mimeType: video.mimeType },
          progress,
        };
      }),
    )
  ).filter((v): v is NonNullable<typeof v> => v !== null);

  res.json({
    empty: false,
    hero: {
      title: hero.name,
      badge: "Recently Added",
      playHref: heroPlayHref,
      infoHref: heroInfoHref ?? null,
      thumbnailPath: heroThumbnailPath,
    },
    continueWatching: continueWatchingItems,
    folders,
    videos,
  });
});

router.get("/folder/:folderId", requireAuth, async (req, res) => {
  const { folderId } = req.params;
  const folder = await getCatalogFolder(folderId);
  if (!folder || folder.status !== "published") {
    res.status(404).json({ error: "Not found" });
    return;
  }

  const backHref =
    folder.parentFolderId === env.driveRootFolderId ? "/" : `/folder/${folder.parentFolderId}`;

  // Case 1: no subfolders — just a flat grid of whatever videos/folders sit directly inside.
  const childFolderDocs = await listPublishedFoldersByParent(folderId);
  if (childFolderDocs.length === 0) {
    const [videoDocs, progressList] = await Promise.all([
      listPublishedVideosByParent(folderId),
      getProgressForFolder(req.authUser!.userId, folderId),
    ]);
    const items = naturalSort([
      ...childFolderDocs.map((f) => ({
        kind: "folder" as const,
        id: f._id,
        name: f.title!,
        thumbnailFileId: f.thumbnailFileId,
      })),
      ...videoDocs.map((v) => ({ kind: "video" as const, id: v._id, name: v.title!, mimeType: v.mimeType })),
    ]);
    res.json({
      mode: "flat",
      folderId,
      title: folder.title,
      backHref,
      thumbnailFileId: folder.thumbnailFileId,
      items,
      progress: progressList,
    });
    return;
  }

  // Case 2: subfolders are seasons — thumbnail, season dropdown, that season's videos, and any
  // videos sitting directly in the series folder itself shown separately as "Specials" (they don't
  // change when the season switches).
  const seasons = naturalSort(childFolderDocs.map((f) => ({ id: f._id, name: f.title! })));
  const requestedSeasonId = typeof req.query.season === "string" ? req.query.season : undefined;
  const selectedSeasonId =
    requestedSeasonId && seasons.some((s) => s.id === requestedSeasonId)
      ? requestedSeasonId
      : seasons[0].id;

  const [seasonVideoDocs, specialVideoDocs, progressList] = await Promise.all([
    listPublishedVideosByParent(selectedSeasonId),
    listPublishedVideosByParent(folderId),
    getProgressForFolder(req.authUser!.userId, folderId),
  ]);

  const seasonItems = orderEpisodes(
    seasonVideoDocs.map((v) => ({
      kind: "video" as const,
      id: v._id,
      name: v.title!,
      mimeType: v.mimeType,
      episodeOrder: v.episodeOrder,
    })),
  );
  const specialItems = orderEpisodes(
    specialVideoDocs.map((v) => ({
      kind: "video" as const,
      id: v._id,
      name: v.title!,
      mimeType: v.mimeType,
      episodeOrder: v.episodeOrder,
    })),
  );

  res.json({
    mode: "series",
    folderId,
    title: folder.title,
    backHref,
    thumbnailFileId: folder.thumbnailFileId,
    seasons,
    selectedSeasonId,
    seasonItems,
    specialItems,
    progress: progressList,
  });
});

router.get("/video/:fileId", requireAuth, async (req, res) => {
  const { fileId } = req.params;
  const video = await getCatalogVideo(fileId);

  // "pending" (no title yet) is never watchable — preview happens on the curation page instead.
  // "curated" (title set, not yet published) is admin-only, so you can check timing/subtitles
  // before making it visible. "published" is open to any signed-in viewer.
  const isAdmin = isAdminEmail(req.authUser!.email);
  if (!video || video.status === "pending" || (video.status === "curated" && !isAdmin)) {
    res.status(404).json({ error: "Not found" });
    return;
  }

  const parentFolderId = video.parentFolderId;
  const isStandalone = parentFolderId === env.driveRootFolderId;

  let nextFileId: string | null = null;
  let previousFileId: string | null = null;
  let episodes: { id: string; name: string }[] = [];
  let progressByFileId: Record<string, { positionSeconds: number; completed: boolean }> = {};

  if (!isStandalone) {
    const [episodeDocs, progressList] = await Promise.all([
      listPublishedVideosByParent(parentFolderId),
      getProgressForFolder(req.authUser!.userId, parentFolderId),
    ]);
    const episodeList = orderEpisodes(
      episodeDocs.map((e) => ({ id: e._id, name: e.title!, episodeOrder: e.episodeOrder })),
    );
    const index = episodeList.findIndex((e) => e.id === fileId);
    nextFileId = index >= 0 ? (episodeList[index + 1]?.id ?? null) : null;
    previousFileId = index >= 0 ? (episodeList[index - 1]?.id ?? null) : null;
    episodes = episodeList;
    progressByFileId = Object.fromEntries(
      progressList.map((p) => [p.fileId, { positionSeconds: p.positionSeconds, completed: p.completed }]),
    );
  }

  const progress = await getProgress(req.authUser!.userId, fileId);
  const backHref = isStandalone ? "/" : `/folder/${parentFolderId}`;
  // "native": browser <video> plays it directly. "raw": MKV — not web-playable at all, only the
  // desktop/mobile app's native player handles it (see isMkv in lib/drive.ts and the `raw` query
  // param on routes/stream.ts). "restart": anything else, remuxed live on every seek.
  const seekMode = isNativelyPlayable(video.mimeType) ? "native" : isMkv(video.mimeType) ? "raw" : "restart";

  const subtitleSets = await listSubtitleSetsByIds(video.subtitleSetIds ?? []);
  // Re-index sequentially — track `index` is only meaningful within the source it came from, and
  // multiple merged sets could otherwise collide (used as the player's <track> key).
  const subtitles: ConvertedSubtitle[] = subtitleSets
    .flatMap((s) => s.subtitles)
    .map((s, i) => ({ ...s, index: i }));

  res.json({
    fileId,
    title: video.title,
    backHref,
    parentFolderId,
    nextFileId,
    previousFileId,
    episodes,
    progressByFileId,
    initialPositionSeconds: progress?.positionSeconds ?? 0,
    initialCompleted: progress?.completed ?? false,
    // Remembered track choices (see lib/progress.ts) — undefined/absent on `progress` itself
    // (never saved) collapses to null here, meaning "no preference, use default behavior".
    initialSubtitleSource: progress?.subtitleSource ?? null,
    initialSubtitleIndex: progress?.subtitleIndex ?? null,
    initialAudioLanguage: progress?.audioLanguage ?? null,
    initialAudioTitle: progress?.audioTitle ?? null,
    seekMode,
    durationSeconds: video.durationSeconds,
    subtitles,
    introStart: video.introStart,
    introEnd: video.introEnd,
    outroStart: video.outroStart,
  });
});

export default router;
